File: docs/current_docs/manuals/developer/python/snippets/services/service-lifecycle-2/main.py
from dagger import dag, function, object_type


@object_type
class MyModule:
    @function
    async def redis_service(self) -> str:
        # create Redis service container
        redis_srv = dag.container().from_("redis").with_exposed_port(6379).as_service()

        # create Redis client container
        redis_cli = (
            dag.container()
            .from_("redis")
            .with_service_binding("redis-srv", redis_srv)
            .with_entrypoint(["redis-cli", "-h", "redis-srv"])
        )

        # set value
        setter = await redis_cli.with_exec(["set", "foo", "abc"]).stdout()

        # get value
        getter = await redis_cli.with_exec(["get", "foo"]).stdout()

        return setter + getter

File: docs/current_docs/manuals/developer/go/snippets/services/service-lifecycle-2/main.go
package main

import (
	"context"
)

type MyModule struct{}

// creates Redis service and client
func (m *MyModule) RedisService(ctx context.Context) (string, error) {
	redisSrv := dag.Container().
		From("redis").
		WithExposedPort(6379).
		AsService()

	// create Redis client container
	redisCLI := dag.Container().
		From("redis").
		WithServiceBinding("redis-srv", redisSrv).
		WithEntrypoint([]string{"redis-cli", "-h", "redis-srv"})

	// set value
	setter, err := redisCLI.
		WithExec([]string{"set", "foo", "abc"}).
		Stdout(ctx)
	if err != nil {
		return "", err
	}

	// get value
	getter, err := redisCLI.
		WithExec([]string{"get", "foo"}).
		Stdout(ctx)
	if err != nil {
		return "", err
	}

	return setter + getter, nil
}
